plugins/user.py

```diff
@@ -0,0 +1,220 @@
+        """
+        username = message.sender.nick
+        if verify_twofactor(self,username, token):
+            self.reply(message, "You are authenticated, {}".format(username)) 
+        else:
+            self.reply(message, "That's not correct,{}".format(username))
+
+
+    @respond_to("^twofactor me")
+    def create_user_twofactor(self, message):
+        """
+        twofactor me: set up twofactor authentication for your user
+        """
+        username = message.sender.nick
+        user_twofactor = self.load("user_twofactor", {})
+        if username not in user_twofactor:
+            user_twofactor[username] = pyotp.random_base32()
+            message['type'] = 'chat'
+            self.reply(message, self.generate_and_upload_QR(user_twofactor[username], username))
+            self.save("user_twofactor", user_twofactor)
+            self.reply(message, "say 'twofactor verify <token>' to check your twofactor verification")
+        else:
+            message['type'] = 'chat'
+            self.reply(message, "twofactor is already set up!") 
+
+    @respond_to("^twofactor remove (?P<username>\w*)")
+    def remove_user_twofactor(self, message, username):
+        """
+        twofactor remove [username]: remove [username]'s twofactor authentication (requires 'admin' permission)
+        """
+        if check_user_permission(self, message.sender.nick, 'admin'):  
+            user_twofactor = self.load("user_twofactor", {})
+            if username in user_twofactor:
+                del user_twofactor[username]
+                self.reply(message, "twofactor is removed for {}".format(username)) 
+                self.save("user_twofactor", user_twofactor)
+            else:
+                self.reply(message, "twofactor was not set for {}".format(username)) 
+        else:
+            self.say("@{}: you don't have admin permission".format(message.sender.nick), message=message) 
+
+    @respond_to("^what can (?P<username>\w*) do")
+    def show_user_permission(self, message, username):
+        """
+        what can [username] do?: get someone's permissions
+        """
+        if 'I' == username:
+            username = message.sender.nick
+        user_permissions = self.load("user_permissions", {})
+        try:
+            self.say("@{}: {}'s permissions are: {}".format(message.sender.nick, username, ', '.join(user_permissions[username])), message=message) 
+        except KeyError:
+            self.say("@{}: I don't know who {} is. (no permissions)".format(message.sender.nick, username), message=message)
+
+
+    @respond_to("^who can(?P<permissions>( \w*)*)")
+    def find_user_permission(self, message, permissions):
+        """
+        who can [permission]?: find the list of people with a permission
+        """
+        user_permissions = self.load("user_permissions", {})
+        for permission in permissions.split():
+            userlist = []
```

Comment:
Great place to use a list comprehension! `userlist = [user for user, user_perm in user_permissions.items() for permission in permissions.split() if permission in user_perm]`